typed-protocols/src/Network/TypedProtocol/Core.hs

```diff
@@ -0,0 +1,355 @@
+{-# LANGUAGE GADTs #-}
+{-# LANGUAGE TypeFamilies #-}
+{-# LANGUAGE RankNTypes #-}
+{-# LANGUAGE PolyKinds #-}
+{-# LANGUAGE DataKinds #-}
+
+
+-- | This module defines the core of the typed protocol framework.
+--
+
+module Network.TypedProtocol.Core (
+
+  -- * Introduction
+  -- $intro
+
+
+  -- * Defining protocols
+  -- $defining
+  Protocol(..),
+
+  -- * Engaging in protocols
+  -- $using
+  PeerKind(..),
+  PeerHasAgency(..),
+  WeHaveAgency,
+  TheyHaveAgency,
+  Peer(..),
+
+  -- * Protocol proofs and tests
+  -- $tests
+  ) where
+
+-- $intro
+-- A typed protocol between two peers is defined via a state machine: a
+-- collection of protocol states and protocol messages which are transitions
+-- between those states.
+--
+-- Start from the idea that a protocol is some language of messages sent
+-- between two peers. To specify a protocol is to describe what possible
+-- sequences of messages are valid. One simple but still relatively expressive
+-- way to do this is via a state machine: starting from some initial state,
+-- all possible paths through the state machine gives the set of valid protocol
+-- traces. This then dictates what a peer participating in a protocol may
+-- produce and what it must accept.
+--
+-- In this style we have a fixed number of states and in each state there is
+-- some number of valid messages that move us on to the next state. This can be
+-- illustrated as a graph, which can be a helpful form of documentation.
+--
+-- We further constrain this idea by saying that the two peers will use the
+-- same state machine and change states in lock-step by sending\/receiving
+-- messages. In this approach, for each protocol state, the description
+-- dictates which peer has the agency to choose to send a message, while
+-- correspondingly the other must be prepared to receive the message.
+--
+-- The views of the two peers are dual. In each state one peer can send any
+-- message that is valid for the current protocol state while the other
+-- must be prepared to receive any valid message for current protocol state.
+--
+-- We can also have terminal protocol states in which neither peer has agency.
+--
+-- So part of the protocol description is to label each protocol state with
+-- the peer that has the agency in that state, or none for terminal states.
+-- We use the labels \"client\" and \"server\" for the two peers, but they are
+-- in fact symmetric.
+
+
+-- $defining
+--
+-- The 'Protocol' type class bundles up all the requirements for a typed
+-- protocol, which are in fact all type level constructs. Defining a new
+-- protocol and making it an instance of the 'Protocol' class requires the
+-- following language extensions:
+--
+-- > {-# LANGUAGE GADTs, TypeFamilies, DataKinds #-}
+--
+-- The type class itself is indexed on a protocol \"tag\" type. This type
+-- does double duty as the /kind/ of the /types/ of the protocol states.
```

Comment:
That is right. It is a limitation. I tried adding another type family so that the tag and the protocol states could be separate, but I ran into a ghc bug that could only be solved with an ugly workaround. This could be improved if/when that ghc issue is resolved.

Reply:
Is it the fact that data kind constructors are not promoted?

Reply:
It's that we have to have the type family for the state types, and then use that for the kind annotations in the message and state partiton definitons. This ends up as one big recusrive binding group that ghc does the kind & type inference on. For some reason it doesn't do the dependency analysis right and then gets confused. If one breaks up the recusive binding group by splitting into two modules (or equiv hack with TH decls) then it works. But this is pretty ugly.